src/libraries/System.Private.CoreLib/src/System/IO/StreamReader.cs

```diff
@@ -879,63 +877,89 @@ private int ReadBuffer(Span<char> userBuffer, out bool readToUserBuffer)
                 {
-                    char ch = tmpCharBuffer[i];
-
                     // Note the following common line feed chars:
                     // \n - UNIX   \r\n - DOS   \r - Mac
-                    if (ch == '\r' || ch == '\n')
+                    int i = _charBuffer.AsSpan(_charPos, _charLen - _charPos).IndexOfAny('\r', '\n');
+                    if (i >= 0)
                     {
-                        string s;
-
-                        if (sb != null)
+                        string? s;
+                        if (rentedArray != null)
                         {
-                            sb.Append(tmpCharBuffer, tmpCharPos, i - tmpCharPos);
-                            s = sb.ToString();
+                            Append(ref rentedArray, lastWrittenIndex, _charBuffer, _charPos, i);
+                            lastWrittenIndex += i;
+                            s = new string(rentedArray!, 0, lastWrittenIndex);
                         }
                         else
                         {
-                            s = new string(tmpCharBuffer, tmpCharPos, i - tmpCharPos);
+                            s = new string(_charBuffer, _charPos, i);
                         }
-
-                        _charPos = tmpCharPos = i + 1;
-
-                        if (ch == '\r' && (tmpCharPos < tmpCharLen || (await ReadBufferAsync(cancellationToken).ConfigureAwait(false)) > 0))
+                        char ch = _charBuffer[_charPos + i];
+                        _charPos += i + 1;
+                        if (ch == '\r' && (_charPos < _charLen || (await ReadBufferAsync(cancellationToken).ConfigureAwait(false)) > 0))
                         {
-                            tmpCharPos = _charPos;
-                            if (_charBuffer[tmpCharPos] == '\n')
+                            if (_charBuffer[_charPos] == '\n')
                             {
-                                _charPos = ++tmpCharPos;
+                                _charPos++;
                             }
                         }
 
                         return s;
                     }
 
-                    i++;
-                } while (i < tmpCharLen);
+                    i = _charLen - _charPos;
+                    if (i < 0) break; // a hack for System.Globalization.Tests (to check on CI if it fixes the problem)
 
-                i = tmpCharLen - tmpCharPos;
-                sb ??= new StringBuilder(i + 80);
-                sb.Append(tmpCharBuffer, tmpCharPos, i);
-            } while (await ReadBufferAsync(cancellationToken).ConfigureAwait(false) > 0);
-
-            return sb.ToString();
+                    Append(ref rentedArray, lastWrittenIndex, _charBuffer, _charPos, i);
+                    lastWrittenIndex += i;
+                } while (await ReadBufferAsync(cancellationToken).ConfigureAwait(false) > 0);
+                return new string(rentedArray!, 0, lastWrittenIndex);
+            }
+            finally
```

Comment:
As a general rule, I don't believe we generally bother with a try/finally simply to return a pooled array. Hopefully exceptions are exceptional, and in this case failing to return an array is a small issue.
The key rules are to return it in the normal case, to only return it once, and to not use it after returning.

Reply:
If something goes wrong then we wouldn’t be able to return to the shared pool. I don’t know but it seems like an issue to me, that’s why have a try/finally block. Would do you suggest?

Reply:
You're correct, but if something goes wrong then we have a bigger problem than wasting an allocation. And so we prefer the simpler code.

Reply:
Do we really have a problem if a cancellation token triggered on an IO operation (ReadBufferAsync)? That's like a valid case where we wouldn't return to the shared pool (if we omit the finally block) but continue with the normal execution.  

I feel like I don't understand the suggestion or what you mean, sorry :(

Reply:
The situation we're trying to avoid is something holding on to the buffer after the exception is thrown and returning into the pool prematurely. Lets say for example ReadBufferAsync failed in a way where the buffer was still being used by a background thread (a rogue Task.Run). This finally would run, return to the pool, then another piece of code would rent it and it would still be in used by the rogue Task.Run. To count this out, we don't bother returning to the pool in exceptional cases.

Reply:
Now I get it. But still `ReadBufferAsync` doesn't use our rented array and it's being rented and used only in this place so no rogue task could use it since it doesn't get forwarded to any of the functions down to call-stack.

Reply:
My point is that it makes the code cleaner to not have the try/ finally. ArrayPool is all about stats, we’re assuming there will be far far more successes here than exceptions.
Also try/finally are not necessarily free: they may impede some codegen optimizations (eg inlining, although perhaps not relevant here)